Demo/Common/IntegrationExamples/Components/CardComponentAdvancedFlowExample.swift

```diff
@@ -0,0 +1,200 @@
+                                                                           target: self,
+                                                                           action: #selector(cancelPressed))
+        presenter?.present(viewController: navigation, completion: nil)
+    }
+
+    @objc private func cancelPressed() {
+        cardComponent?.cancelIfNeeded()
+        presenter?.dismiss(completion: nil)
+    }
+
+    // MARK: - Payment response handling
+
+    private func paymentResponseHandler(result: Result<PaymentsResponse, Error>) {
+        switch result {
+        case let .success(response):
+            if let action = response.action {
+                adyenActionComponent.handle(action)
+            } else {
+                finish(with: response)
+            }
+        case let .failure(error):
+            finish(with: error)
+        }
+    }
+
+    internal func finish(with result: PaymentsResponse) {
+        let success = result.resultCode == .authorised || result.resultCode == .received || result.resultCode == .pending
+        let message = "\(result.resultCode.rawValue) \(result.amount?.formatted ?? "")"
+        finalize(success, message)
+    }
+
+    internal func finish(with error: Error) {
+        let message: String
+        if let componentError = (error as? ComponentError), componentError == ComponentError.cancelled {
+            message = "Cancelled"
+        } else {
+            message = error.localizedDescription
+        }
+        finalize(false, message)
+    }
+
+    private func finalize(_ success: Bool, _ message: String) {
+        cardComponent?.finalizeIfNeeded(with: success) { [weak self] in
+            guard let self = self else { return }
+            self.dismissAndShowAlert(success, message)
+        }
+    }
+
+    internal func dismissAndShowAlert(_ success: Bool, _ message: String) {
+        presenter?.dismiss {
+            // Payment is processed. Add your code here.
+            let title = success ? "Success" : "Error"
+            self.presenter?.presentAlert(withTitle: title, message: message)
+        }
+    }
+
+}
+
+extension CardComponentAdvancedFlowExample: CardComponentDelegate {
+
+    func didSubmit(lastFour: String, finalBIN: String, component: CardComponent) {
+        print("Card used: **** **** **** \(lastFour)")
+        print("Final BIN: \(finalBIN)")
+    }
+
+    internal func didChangeBIN(_ value: String, component: CardComponent) {
+        print("Current BIN: \(value)")
+    }
+
+    internal func didChangeCardBrand(_ value: [CardBrand]?, component: CardComponent) {
+        print("Current card type: \((value ?? []).reduce("") { "\($0), \($1)" })")
+    }
+}
+
+extension CardComponentAdvancedFlowExample: PaymentComponentDelegate {
+
+    internal func didSubmit(_ data: PaymentComponentData, from component: PaymentComponent) {
+        let request = PaymentsRequest(data: data)
+        apiClient.perform(request, completionHandler: paymentResponseHandler)
```

Comment:
Same [weak self] with the previous ones :)

Reply:
e7c03971d28f7f4fb10a23fe39580f2dacb88ac4